### PR TITLE
python312Packages.crewai: init at 0.80.0

### PR DESCRIPTION
**Pull request by AJChapman**
Added these Python modules:

- crewai,
- crewai-tools,
- embedchain,
- google-cloud-aiplatform,
- langchain-cohere,
- langchain-experimental,
- mem0ai,
- pyright,
- agentops

This was all in service of fulfilling CrewAI's transitive dependencies.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -12273,6 +12289,8 @@ self: super: with self; {
 
   pyric = callPackage ../development/python-modules/pyric { };
 
+  pyright = callPackage ../development/python-modules/pyright { };
```
Comment (pineapplehunter) [inline pkgs/top-level/python-packages.nix:12292]: Adding pyright to python packages seems to be causing many packages to rebuild.
It looks like crewai-tools only depends on it for development and is not imported anywhere. I think you could remove pyright and remove the dependency from crewai-tools using something like this:

```nix
pythonRemoveDeps = [
  "pyright"
];
```

Additional resource: https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonrelaxdepshook-using-pythonrelaxdepshook

File: pkgs/development/python-modules/crewai-tools/default.nix [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      repo = "crewAI-tools";
+      rev = "afa7e6a"; # "refs/tags/${version}";
+      hash = "sha256-sxb6nBVoTTOAOs11uwDbtgH9cJ7yc7gwqxhoftgdVMU=";
+    };
+
+    build-system = [ hatchling ];
+
+    nativeBuildInputs =
+      [
+      ];
```
Comment (pineapplehunter) [inline pkgs/development/python-modules/crewai-tools/default.nix:47]:
```suggestion
```

**Review by pineapplehunter — CHANGES_REQUESTED** [review]

**Comment by drupol** [comment]
Can you please update the PR title according to the contributor guide ? See others PRs to get an idea.

File: pkgs/development/python-modules/agentops/default.nix [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      packaging
+      termcolor
+      pyyaml
+    ];
+
+    doCheck = true;
+
+    pythonImportsCheck = [ "agentops" ];
+
+    meta = with lib; {
```
Comment (drupol) [inline pkgs/development/python-modules/agentops/default.nix:56]: We try to avoid `with` statements in front of attribute sets. Same remark for the rest of the files.

File: pkgs/development/python-modules/crewai-tools/default.nix [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  pytube,
+  requests,
+  beautifulsoup4,
+  selenium,
+  docx2txt,
+  docker,
+  embedchain,
+}:
+
+let
```
Comment (drupol) [inline pkgs/development/python-modules/crewai-tools/default.nix:27]: The `let...in` clause is not needed here, I would recommend using `rec` here, and the rest of the files too.
I usually do not advocate for using it, and `buildPythonPackage` doesn't support `finalAttrs` yet.

File: pkgs/development/python-modules/agentops/default.nix (hunk 3 above)
Comment (AJChapman) [reply pkgs/development/python-modules/agentops/default.nix:56]: I can change it, though it seems to be a very common idiom within the python packages at least.

File: pkgs/development/python-modules/crewai-tools/default.nix (hunk 4 above)
Comment (AJChapman) [reply pkgs/development/python-modules/crewai-tools/default.nix:27]: I can remove the unnecessary `let...in` clauses; they were just in imitation of existing packages.

I don't quite understand the comment about `rec`. You *do* recommend it, but usually do not? I am using it so I can reuse the `version` attribute in the `rev` attribute of `fetchFromGitHub`. I can remove at least one instance though.

File: pkgs/development/python-modules/embedchain/default.nix [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+buildPythonPackage {
+  pname = "embedchain";
+  inherit version;
+  pyproject = true;
+
+  disabled = pythonOlder "3.9";
+
+  nativeBuildInputs = [ poetry-core ];
+
+  src = "${repo}/embedchain";
```
Comment (drupol) [inline pkgs/development/python-modules/embedchain/default.nix:54]: I think there's a better way to do that, I don't remember it at the moment, maybe @GaetanLepage knows ?

File: pkgs/development/python-modules/crewai/default.nix [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  version = "0.80.0";
+  pyproject = true;
+  disabled = pythonOlder "3.10";
+  src = fetchFromGitHub {
+    owner = "crewAIInc";
+    repo = "crewAI";
+    rev = "refs/tags/${version}";
+    hash = "sha256-MVWNjV2S2l96tEeaH+3sKftUVRDSkx0dLpfMbWsRRdc=";
+  };
+  passthru.updateScript = gitUpdater { };
```
Comment (drupol) [inline pkgs/development/python-modules/crewai/default.nix:51]: This file is hard to follow, can you please group attributes related to testing together and move it just before the meta ? Also, add a few blank lines to make them easier to read.

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (pineapplehunter) [reply pkgs/top-level/python-packages.nix:12292]: Would you remove `pkgs/development/python-modules/pyright/default.nix` too?

**Review by pineapplehunter — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/embedchain/default.nix (hunk 5 above)
Comment (sarahec) [reply pkgs/development/python-modules/embedchain/default.nix:54]: It's `build-system = [ poetry-core ];` instead.

File: pkgs/development/python-modules/agentops/default.nix [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    psutil
+    packaging
+    termcolor
+    pyyaml
+  ];
+  doCheck = true;
+  pythonImportsCheck = [ "agentops" ];
+  meta = {
+    description = "AgentOps helps developers build, evaluate, and monitor AI agents.";
+    downloadPage = "https://github.com/AgentOps-AI/agentops/releases";
```
Comment (sarahec) [inline pkgs/development/python-modules/agentops/default.nix:47]:
```suggestion
    changelog = "https://github.com/AgentOps-AI/agentops/releases/tag/${version}";
```

Please edit all `downloadPage` entries to be like this.

File: pkgs/development/python-modules/embedchain/default.nix [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  };
+in
+buildPythonPackage {
+  pname = "embedchain";
+  inherit version;
+  pyproject = true;
+
+  disabled = pythonOlder "3.9";
+
+  nativeBuildInputs = [ poetry-core ];
```
Comment (sarahec) [inline pkgs/development/python-modules/embedchain/default.nix:52]:
```suggestion
  build-system = [ poetry-core ];
```

File: pkgs/development/python-modules/embedchain/default.nix (hunk 5 above)
Comment (sarahec) [reply pkgs/development/python-modules/embedchain/default.nix:54]: Switch `repo` to `src` (i.e. move it back into `buildPythonPackage`) then use `sourceRoot = "${src.name}/embedchain";`

File: pkgs/development/python-modules/embedchain/default.nix [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    export EMBEDCHAIN_CONFIG_DIR=$(mktemp -d)
+    export MEM0_DIR=$(mktemp -d)
+  '';
+
+  pythonImportsCheck = [ "embedchain" ];
+
+  meta = {
+    description = "Embedchain is a RAG framework to create data pipelines. It loads, indexes, retrieves and syncs all the data.";
+    downloadPage = "https://github.com/mem0ai/mem0/releases";
+    homepage = "https://python.langchain.com/docs/integrations/retrievers/embedchain/";
```
Comment (sarahec) [inline pkgs/development/python-modules/embedchain/default.nix:101]:
```suggestion
    homepage = "https://github.com/mem0ai/mem0/tree/main/embedchain";
```

File: pkgs/development/python-modules/agentops/default.nix (hunk 7 above)
Comment (AJChapman) [reply pkgs/development/python-modules/agentops/default.nix:47]: You've changed `downloadPage` to `changelog`. Was that deliberate? I'm going to assume not. Will change the `downloadPage` entries as you suggest without changing them to be `changelog` entries.

File: pkgs/development/python-modules/pyright/default.nix [hunk 10]
```diff
@@ -0,0 +1,54 @@
+{
```
Comment (pineapplehunter) [inline pkgs/development/python-modules/pyright/default.nix:1]: I don't think this package is needed to build crewai, and it is causing many packages to rebuild.
Could you remove this package?

**Review by pineapplehunter — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/pyright/default.nix (hunk 10 above)
Comment (AJChapman) [reply pkgs/development/python-modules/pyright/default.nix]: Ok, removed pyright.

Comment (sarahec) [reply pkgs/development/python-modules/pyright/default.nix]: You also need to remove it as a dependency

File: pkgs/development/python-modules/google-cloud-aiplatform/default.nix [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+    packaging
+    google-cloud-storage
+    google-cloud-bigquery
+    google-cloud-resource-manager
+    shapely
+  ] ++ google-api-core.optional-dependencies.grpc;
+
+  optional-dependencies = rec {
+
+    profiler-deps = [
```
Comment (natsukium) [inline pkgs/development/python-modules/google-cloud-aiplatform/default.nix:96]:
```suggestion
    cloud-profiler = [
```
Please correspond to the key in `extras_require`.
https://github.com/googleapis/python-aiplatform/blob/0160d4928e6e1a254a2e594216ae67abf83c7a47/setup.py#L257-L280

Same for other optional dependencies.

File: pkgs/development/python-modules/langchain-cohere/default.nix [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  pythonImportsCheck = [ "langchain_cohere" ];
+
+  passthru = {
+    inherit (langchain-core) updateScript;
+  };
+
+  meta = {
+    description = "Cohere integrations with LangChain";
+    downloadPage = "https://github.com/langchain-ai/langchain-cohere/releases/tag/libs%2Fcohere%2Fv${version}";
+    homepage = "https://pypi.org/project/langchain-cohere/";
```
Comment (natsukium) [inline pkgs/development/python-modules/langchain-cohere/default.nix:58]:
```suggestion
    homepage = "https://github.com/langchain-ai/langchain-cohere/tree/main/libs/cohere";
```
We prefer GitHub instead of PyPI. See other langchain modules.

File: pkgs/development/python-modules/mem0ai/default.nix [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+
+  # propagates
+  qdrant-client,
+  pydantic,
+  openai,
+  posthog,
+  pytz,
+  sqlalchemy,
+
+  # runtime dependencies
```
Comment (natsukium) [inline pkgs/development/python-modules/mem0ai/default.nix:20]: What is the difference with propagates (line 12)?

**Review by natsukium — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/crewai-tools/default.nix [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+  pyproject = true;
+
+  disabled = pythonOlder "3.10";
+
+  src = fetchFromGitHub {
+    owner = "crewAIInc";
+    repo = "crewAI-tools";
+    # Note: There are no tagged releases yet so we just use the latest revision
+    rev = "03f77f0"; # "refs/tags/${version}";
```
[the remark this answers is not in the document]
Comment (AJChapman) [reply pkgs/development/python-modules/crewai-tools/default.nix:38]: I have now asked upstream to add tags: https://github.com/crewAIInc/crewAI-tools/issues/190

File: pkgs/development/python-modules/mem0ai/default.nix (hunk 13 above)
Comment (AJChapman) [reply pkgs/development/python-modules/mem0ai/default.nix:20]: These are things that nixpkgs' runtime checks finds are dependencies, but which mem0 doesn't list as dependencies.

**Comment by sarahec** [comment]
Fixes #296472

File: pkgs/development/python-modules/agentops/default.nix [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  version = "0.3.18";
+
+  pyproject = true;
+
+  disabled = pythonOlder "3.7";
+
+  src = fetchFromGitHub {
+    owner = "AgentOps-AI";
+    repo = "agentops";
+    rev = "refs/tags/${version}";
```
Comment (drupol) [inline pkgs/development/python-modules/agentops/default.nix:29]:
```suggestion
    tag = version;
```

File: pkgs/development/python-modules/agentops/default.nix [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    termcolor
+    pyyaml
+  ];
+
+  pythonRelaxDeps = [
+    "psutil"
+    "packaging"
+  ];
+
+  build-system = [ setuptools ];
```
Comment (drupol) [inline pkgs/development/python-modules/agentops/default.nix:46]: Usually the build-system comes before the dependencies

File: pkgs/development/python-modules/crewai/default.nix [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+  version = "0.80.0";
+
+  pyproject = true;
+
+  disabled = pythonOlder "3.10";
+
+  src = fetchFromGitHub {
+    owner = "crewAIInc";
+    repo = "crewAI";
+    rev = "refs/tags/${version}";
```
Comment (drupol) [inline pkgs/development/python-modules/crewai/default.nix:51]:
```suggestion
    tag = version;
```

**Comment by adam248** [comment]
has this PR stalled? crewai version is now @  ~v0.121.1 maybe this should be redone with that version?

**Comment by AdrienLemaire** [comment]
With the new [ACP course on DeepLearning](https://learn.deeplearning.ai/courses/acp-agent-communication-protocol), I would have loved to test out CrewAI on NixOS.

**Comment by sarahec** [comment]
Here's how I'd move forward:
1. Review each of the packages this adds (other than crewai). Some of them now exist and can be removed from this PR. The remaining ones should be filed as individual PRs (`python3Packages.foo: init at 3.14.159`)
2. Rebase this PR onto `master` and deal with any conflicts

Alternately, close this and start a series of new PRs (`init` any missing dependencies in single PRs, then this as a single PR.) Small PRs tend to be reviewed quicker than big ones.
